src/processing/script/destructuring.ts

```diff
@@ -0,0 +1,143 @@
+import {
+    Expression,
+    ObjectPattern,
+    Pattern,
+    ArrayPattern,
+    AssignmentPattern,
+    AssignmentProperty,
+    Identifier,
+    CallExpression
+} from 'estree';
+import { Syntax } from 'esotope-hammerhead';
+import {
+    createArrayExpression,
+    createBinaryExpression,
+    createConditionalExpression,
+    createIdentifier,
+    createMemberExpression,
+    createSimpleCallExpression,
+    createSimpleLiteral,
+    createUndefined
+} from './node-builder';
+import TempVariables from './transformers/temp-variables';
+import INSTRUCTION from './instruction';
+
+type NodeBuilder = (pattern: Pattern, value: Expression, isTemp?: boolean) => void;
+
+function processObjectProperty (prop: AssignmentProperty, temp: Identifier, build: NodeBuilder, baseTempName: string) {
+    const pattern  = prop.value;
+    const computed = prop.computed || prop.key.type === Syntax.Literal;
+    const value    = createMemberExpression(temp, prop.key, computed);
+
+    process(pattern, value, build, baseTempName);
+}
+
+function createObjectRest (tempIdentifier: Identifier, keys: Expression[]) {
+    const restObjectIdentifier = createIdentifier(INSTRUCTION.restObject);
+
+    return createSimpleCallExpression(restObjectIdentifier, [tempIdentifier, createArrayExpression(keys)]);
+}
+
+function createRestArray (array: Identifier, startIndex: number): CallExpression {
+    const restArrayIdentifier = createIdentifier(INSTRUCTION.restArray);
+
+    return createSimpleCallExpression(restArrayIdentifier, [array, createSimpleLiteral(startIndex)]);
+}
+
+function processObjectPattern (pattern: ObjectPattern, value: Expression, build: NodeBuilder, baseTempName: string) {
+    if (!baseTempName)
+        baseTempName = TempVariables.generateName(baseTempName);
+
+    const properties     = pattern.properties;
+    // @ts-ignore
+    const hasRest        = properties[properties.length - 1].type === Syntax.RestElement;
```

Comment:
Is it guaranteed that length is always greater than 1? Empty destructuring is an edge case, but it is a valid expression:
```js
console.log({}={a:5})
```